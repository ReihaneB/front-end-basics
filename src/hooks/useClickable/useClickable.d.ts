import React from 'react';

interface useClickableProps {
  onClick?: (e: React.MouseEvent | React.TouchEvent) => void;
  onMouseDown?: (e: React.MouseEvent) => void;
  onMouseEnter?: (e: React.MouseEvent) => void;
  disabled?: boolean;
  disableDebounce?: boolean;
}
