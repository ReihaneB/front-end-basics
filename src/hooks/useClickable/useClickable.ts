import React, { useCallback, useMemo } from 'react';

import { useDebounceCallback } from '@/hooks/useDebounceCallback/useDebounceCallback';

import { useClickableProps } from './useClickable.d';

export function useClickable({
  onClick = () => {},
  onMouseDown = () => {},
  onMouseEnter = () => {},
  disabled = false,
  disableDebounce = false,
}: useClickableProps) {
  const handleClick = useDebounceCallback({
    callback: onClick,
    waitTime: 300,
    disableDebounce,
  });

  const handleMouseDown = useDebounceCallback({
    callback: onMouseDown,
    waitTime: 300,
    disableDebounce,
  });

  const handleMouseOver = useCallback((e: React.MouseEvent<Element, MouseEvent>) => {
    onMouseEnter?.(e);
  }, [onMouseEnter]);

  return useMemo(() => ({
    eventHandlers: {
      onClick: !disabled ? handleClick : undefined,
      onMouseDown: !disabled ? handleMouseDown : undefined,
      onMouseOver: handleMouseOver,
    },
  }), [
    handleClick,
    handleMouseDown,
    handleMouseOver,
  ]);
}
