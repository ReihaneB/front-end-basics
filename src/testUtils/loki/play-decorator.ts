/* eslint-disable @typescript-eslint/ban-ts-comment */
// @ts-expect-error
import createAsyncCallback from '@loki/create-async-callback';
// @ts-expect-error
import isLokiRunning from '@loki/is-loki-running';
import type {PlayFunction} from '@storybook/csf';
// @ts-expect-error
import type {ReactFramework} from '@storybook/react';

async function waitForDocumentLoaded(): Promise<void> {
  if (document.readyState === 'loading') {
    return new Promise(resolve => {
      document.addEventListener('DOMContentLoaded', () => resolve());
    });
  }

  return Promise.resolve();
}

function enablePointerEvents(): void {
  const styleElement = document.createElement('style');
  document.documentElement.appendChild(styleElement);
  styleElement.sheet?.insertRule('* {pointer-events: auto !important}');
}

export function lokiPlayDecorator(
  target: PlayFunction<ReactFramework>
): PlayFunction<ReactFramework> {
  return async context => {
    const lokiRunning = isLokiRunning();
    const asyncCallback = createAsyncCallback();

    if (lokiRunning) {
      await waitForDocumentLoaded();
      enablePointerEvents();
    }

    try {
      await target(context);
    } finally {
      asyncCallback();
    }
  };
}
