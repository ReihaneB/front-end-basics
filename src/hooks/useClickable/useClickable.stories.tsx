import { useState } from 'react';
import { expect, userEvent, waitFor, within } from '@storybook/test';

import { lokiPlayDecorator } from '@/testUtils/loki/play-decorator';

import { useClickable } from './useClickable';

export default {
  title: 'Hooks/useClickable',
  component: useClickable,
};

function ClickableComponent({
  useClickableProps = {},
}) {
  const [count, setCount] = useState(0);

  const handleClick = () => {
    setCount(prevState => prevState + 1);
  };

  const { eventHandlers } = useClickable({ onClick: handleClick, ...useClickableProps });

  return (
    <div>
      <button
        type="button"
        data-testid="clickable"
        {...eventHandlers}
      >
        Click me
      </button>
      <p
        data-testid="count"
      >
        Clicked
        {' '}
        {count}
        {' '}
        times
      </p>
    </div>
  );
}

export function Playground({...args}) {
  return <ClickableComponent {...args} />;
}

export function ClickOnce() {
  return <ClickableComponent />;
}

ClickOnce.play = lokiPlayDecorator(
  async ({ canvasElement }: { canvasElement: HTMLElement }) => {
    const canvas = within(canvasElement);

    await userEvent.dblClick(canvas.getByTestId('clickable'));

    await waitFor(() => expect(canvas.getByTestId('count')).toHaveTextContent('Clicked 1 times'));
  }
);

export function DisableDebounce() {
  return (
    <ClickableComponent useClickableProps={{ disableDebounce: true }} />
  );
}

DisableDebounce.play = lokiPlayDecorator(
  async ({ canvasElement }: { canvasElement: HTMLElement }) => {
    const canvas = within(canvasElement);

    await userEvent.dblClick(canvas.getByTestId('clickable'));

    expect(canvas.getByTestId('count')).toHaveTextContent('Clicked 2 times');
  }
);

export function Disabled() {
  return (
    <ClickableComponent useClickableProps={{ disabled: true }} />
  );
}

Disabled.play = lokiPlayDecorator(
  async ({ canvasElement }: { canvasElement: HTMLElement }) => {
    const canvas = within(canvasElement);

    await userEvent.click(canvas.getByTestId('clickable'));

    expect(canvas.getByTestId('count')).toHaveTextContent('Clicked 0 times');
  }
);

export function WithOnMouseDown() {
  const [callMouseDown, setCallMouseDown] = useState(0);

  const handleMouseDown = () => {
    setCallMouseDown(prevState => prevState + 1);
  };
  return (
    <div>
      <p>
        onMouseDown called
        {' '}
        {callMouseDown}
        {' '}
        times
      </p>
      <ClickableComponent useClickableProps={{ onMouseDown: handleMouseDown }} />
    </div>
  );
}

WithOnMouseDown.play = lokiPlayDecorator(
  async ({ canvasElement }: { canvasElement: HTMLElement }) => {
    const canvas = within(canvasElement);

    await userEvent.type(canvas.getByTestId('clickable'), '{mousedown}');

    await waitFor(() => expect(canvas.getByTestId('count')).toHaveTextContent('Clicked 1 times'));
  }
);

export function PressEnter() {
  const [callMouseDown, setCallMouseDown] = useState(0);

  const handleMouseDown = () => {
    setCallMouseDown(prevState => prevState + 1);
  };
  return (
    <div>
      <p>
        onMouseDown called
        {' '}
        {callMouseDown}
        {' '}
        times
      </p>
      <ClickableComponent useClickableProps={{ onMouseDown: handleMouseDown }} />
    </div>
  );
}

PressEnter.play = lokiPlayDecorator(
  async ({ canvasElement }: { canvasElement: HTMLElement }) => {
    const canvas = within(canvasElement);

    await userEvent.type(canvas.getByTestId('clickable'), '{enter}');

    await waitFor(() => expect(canvas.getByTestId('count')).toHaveTextContent('Clicked 1 times'));
  }
);

export function PressSpace() {
  const [callMouseDown, setCallMouseDown] = useState(0);

  const handleMouseDown = () => {
    setCallMouseDown(prevState => prevState + 1);
  };
  return (
    <div>
      <p>
        onMouseDown called
        {' '}
        {callMouseDown}
        {' '}
        times
      </p>
      <ClickableComponent useClickableProps={{ onMouseDown: handleMouseDown }} />
    </div>
  );
}

PressSpace.play = lokiPlayDecorator(
  async ({ canvasElement }: { canvasElement: HTMLElement }) => {
    const canvas = within(canvasElement);

    await userEvent.type(canvas.getByTestId('clickable'), '{space}');

    await waitFor(() => expect(canvas.getByTestId('count')).toHaveTextContent('Clicked 1 times'));
  }
);

export function WithOnMouseEnter() {
  const [callMouseEnter, setCallMouseEnter] = useState(0);

  const handleMouseEnter = () => {
    setCallMouseEnter(prevState => prevState + 1);
  };
  return (
    <div>
      <p data-testid="mouse-enter">
        onMouseEnter called
        {' '}
        {callMouseEnter}
        {' '}
        times
      </p>
      <ClickableComponent useClickableProps={{ onMouseEnter: handleMouseEnter }} />
    </div>
  );
}

WithOnMouseEnter.play = lokiPlayDecorator(
  async ({ canvasElement }: { canvasElement: HTMLElement }) => {
    const canvas = within(canvasElement);

    await userEvent.hover(canvas.getByTestId('clickable'));
    await userEvent.hover(canvas.getByTestId('clickable'));
    await userEvent.hover(canvas.getByTestId('clickable'));

    await waitFor(() => expect(canvas.getByTestId('mouse-enter')).toHaveTextContent('onMouseEnter called 3 times'));
    await waitFor(() => expect(canvas.getByTestId('count')).toHaveTextContent('Clicked 0 times'));
  }
);
