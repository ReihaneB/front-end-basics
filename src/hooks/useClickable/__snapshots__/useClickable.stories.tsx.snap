// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hooks/useClickable ClickOnce play-test 1`] = `
<div>
  <button type="button"
          data-testid="clickable"
  >
    Click me
  </button>
  <p data-testid="count">
    Clicked 1 times
  </p>
</div>
`;

exports[`Hooks/useClickable DisableDebounce play-test 1`] = `
<div>
  <button type="button"
          data-testid="clickable"
  >
    Click me
  </button>
  <p data-testid="count">
    Clicked 2 times
  </p>
</div>
`;

exports[`Hooks/useClickable Disabled play-test 1`] = `
<div>
  <button type="button"
          data-testid="clickable"
  >
    Click me
  </button>
  <p data-testid="count">
    Clicked 0 times
  </p>
</div>
`;

exports[`Hooks/useClickable MultipleClicks play-test 1`] = `
<div>
  <button type="button"
          data-testid="clickable"
  >
    Click me
  </button>
  <p data-testid="count">
    Clicked 1 times
  </p>
</div>
`;

exports[`Hooks/useClickable Playground smoke-test 1`] = `
<div>
  <button type="button"
          data-testid="clickable"
  >
    Click me
  </button>
  <p data-testid="count">
    Clicked 0 times
  </p>
</div>
`;

exports[`Hooks/useClickable PressEnter play-test 1`] = `
<div>
  <p>
    onMouseDown called 1 times
  </p>
  <div>
    <button type="button"
            data-testid="clickable"
    >
      Click me
    </button>
    <p data-testid="count">
      Clicked 1 times
    </p>
  </div>
</div>
`;

exports[`Hooks/useClickable PressSpace play-test 1`] = `
<div>
  <p>
    onMouseDown called 1 times
  </p>
  <div>
    <button type="button"
            data-testid="clickable"
    >
      Click me
    </button>
    <p data-testid="count">
      Clicked 1 times
    </p>
  </div>
</div>
`;

exports[`Hooks/useClickable WithOnMouseDown play-test 1`] = `
<div>
  <p>
    onMouseDown called 1 times
  </p>
  <div>
    <button type="button"
            data-testid="clickable"
    >
      Click me
    </button>
    <p data-testid="count">
      Clicked 1 times
    </p>
  </div>
</div>
`;

exports[`Hooks/useClickable WithOnMouseEnter play-test 1`] = `
<div>
  <p data-testid="mouse-enter">
    onMouseEnter called 3 times
  </p>
  <div>
    <button type="button"
            data-testid="clickable"
    >
      Click me
    </button>
    <p data-testid="count">
      Clicked 0 times
    </p>
  </div>
</div>
`;
